OSSMDOC-406: Verify instructions in section "Verifying your certifica… (#41446)
openshift/openshift-docs: modules/ossm-security-cert-manage.adoc

* Install {ProductName} with mutual TLS enabled to configure certificates.
* This example uses the certificates from the link:https://github.com/maistra/istio/tree/maistra-{MaistraVersion}/samples/certs[Maistra repository]. For production, use your own certificates from your certificate authority.
* Deploy the Bookinfo sample application to verify the results with these instructions.

> * Install OpenSSL to verify certificates.

Reviewer: I would change this to "OpenSSL is required to verify certificates" (because it might already be installed, so no need to install it again).

Author: made edit

Add the certificates to {ProductShortName} by following these steps. Save the example certificates from the link:https://github.com/maistra/istio/tree/maistra-{MaistraVersion}/samples/certs[Maistra repository] locally and replace `<path>` with the path to your certificates.

> . Create a secret named `cacert` that includes the input files `ca-cert.pem`, `ca-key.pem`, `root-cert.pem` and `cert-chain.pem`.

Reviewer: Line 22 - This is the second time I've walked through this and missed the line about downloading the sample files.  Can we make step 1 "Save the example certficates....."

Author: edit made

. Verify that the root certificate is the same as the one specified by the administrator. Replace `<path>` with the path to your certificates.
+
[source,terminal]
----
$ openssl x509 -in <path>/root-cert.pem -text -noout > /tmp/root-cert.crt.txt
$ openssl x509 -in ./proxy-cert-3.pem -text -noout > /tmp/pod-root-cert.crt.txt
$ diff -s /tmp/root-cert.crt.txt /tmp/pod-root-cert.crt.txt
Files /tmp/root-cert.crt.txt and /tmp/pod-root-cert.crt.txt are identical
----

. Verify that the CA certificate is the same as the one specified by the administrator. Replace `<path>` with the path to your certificates.
+
[source,terminal]
----
$ openssl x509 -in ./proxy-cert-2.pem -text -noout > /tmp/pod-cert-chain-ca.crt.txt
$ diff -s /tmp/ca-cert.crt.txt /tmp/pod-cert-chain-ca.crt.txt
Files /tmp/ca-cert.crt.txt and /tmp/pod-cert-chain-ca.crt.txt are identical
----

. Verify the certificate chain from the root certificate to the workload certificate. Replace `<path>` with the path to your certificates.
+
[source,terminal]
----
$ openssl verify -CAfile <(cat <path>/ca-cert.pem <path>/root-cert.pem) ./proxy-cert-1.pem
./proxy-cert-1.pem: OK
----

> 

Reviewer 2: Are these blank lines intentional?

Author: Thanks for pointing this out. I submitted a revision to clean it up.

      istiod:
        type: PrivateKey
        privateKey:
          rootCADir: /etc/cacerts
----

> [NOTE]

Reviewer: Curious why you chose to make this a note rather than another step in the procedure?  I probably would have made it a step.  It's fine either way, I'm really just curious, did this feel like it needed more emphasis than a step?

Author: Yes, it was done for emphasis because in similar circumstances customer had issues in the field. They didn't always know when restating pods was essential.

====
After creating/changing/deleting the `cacert` secret, the control plane `istiod` and `gateway` pods must be restarted so the changes go into effect. Use the following command to restart the pods: `oc -n istio-system delete pods -l 'app in (istiod,istio-ingressgateway)'`.

Restart the application pods so that the sidecar containers pick up the changes as well.
====

[id="ossm-cert-manage-verify-cert_{context}"]
== Verifying your certificates

> Use the xref:../../service_mesh/v2x/ossm-create-mesh.adoc#ossm-tutorial-bookinfo-overview_ossm-create-mesh[Bookinfo example application] to verify that the workload certificates are signed by the certificates that were plugged into the CA. This requires you have `openssl` installed on your machine

Reviewer: We can't put cross-references in modules (yet).  So this link needs to revert back to text.

Author: @JStickler I see this link work in my local build, so I must not understand what you mean.

Reviewer: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#links-hyperlinks-and-cross-references 
"Only use xref in assemblies, not modules."
Technically it works, but right now the rule is not to put XREFs in modules.

Reviewer 3: Should we also include the istio-egressgateway pod in line 53 ?
`oc -n istio-system delete pods -l 'app in (istiod,istio-ingressgateway, istio-egressgateway)'`.

> Use the xref:../../service_mesh/v2x/ossm-create-mesh.adoc#ossm-tutorial-bookinfo-overview_ossm-create-mesh[Bookinfo example application] to verify that the workload certificates are signed by the certificates that were plugged into the CA. This requires you have `openssl` installed on your machine

Reviewer: Also, I would call out the OpenSSL requirement as a prerequisite.

Author: Added bullet in prereqs: " Install OpenSSL to verify certificates."

. Extracting certificates from bookinfo workloads
+
Wait about 1 minute for the mTLS policty to take effect before retrieving the certificate chain of bookinfo. If Bookinfo was already deployed when you created the secret above, you need to restart the pods so they pick up the changes (see note above). Replace `<bookinfo-project>` with the project name where bookinfo workload is installed.
+
[source,terminal]
----

> $ # Restart the bookinfo pods if they were already deployed. Not needed if they are deployed after the secret changes above.

Reviewer: Did you mean for these first two lines to be commented out?  
If you did want these lines as comments, I'd move then out of the code block, so that if customers use the copy/paste function, they don't get these unneeded lines.

Author: Broke them out as a note.

$ # oc -n <bookinfo-project> delete pods --all
$
$ sleep 60
$ oc -n <bookinfo-project> exec "$(oc -n <bookinfo-project> get pod -l app=productpage -o jsonpath={.items..metadata.name})" -c istio-proxy -- openssl s_client -showcerts -connect details:9080 > bookinfo-proxy-cert.txt
$ sed -n '/-----BEGIN CERTIFICATE-----/{:start /-----END CERTIFICATE-----/!{N;b start};/.*/p}' bookinfo-proxy-cert.txt > certs.pem
$ awk 'BEGIN {counter=0;} /BEGIN CERT/{counter++} { print > "proxy-cert-" counter ".pem"}' < certs.pem
----
+
You should have three files in your working dir: `proxy-cert-1.pem`, `proxy-cert-2.pem` and `proxy-cert-3.pem`.

. Verify that the root certificate is the same as the one specified by the administrator. Replace `<path>` with the path to your certificates.
+
[source,terminal]
----
$ openssl x509 -in <path>/root-cert.pem -text -noout > /tmp/root-cert.crt.txt
$ openssl x509 -in ./proxy-cert-3.pem -text -noout > /tmp/pod-root-cert.crt.txt
$ diff -s /tmp/root-cert.crt.txt /tmp/pod-root-cert.crt.txt

> Files /tmp/root-cert.crt.txt and /tmp/pod-root-cert.crt.txt are identical

Reviewer: I think line 86 should be outside of the code block as well.

----

. Verify that the CA certificate is the same as the one specified by the administrator. Replace `<path>` with the path to your certificates.
+
[source,terminal]
----
$ openssl x509 -in ./proxy-cert-2.pem -text -noout > /tmp/pod-cert-chain-ca.crt.txt
$ diff -s /tmp/ca-cert.crt.txt /tmp/pod-cert-chain-ca.crt.txt

> Files /tmp/ca-cert.crt.txt and /tmp/pod-cert-chain-ca.crt.txt are identical

Reviewer: I think line 95 should be outside of the code block as well.

      istiod:
        type: PrivateKey
        privateKey:
          rootCADir: /etc/cacerts
----
+
[NOTE]

> ====

Reviewer: Making this a note means that we've lost the feature where there's a button to copy code or commands to the clipboard.  I think this would be a smoother flow if it was formatted something like:

Step 3 - After creating/changing/deleting the cacert secret, the control plane istiod and gateway pods must be restarted so the changes go into effect. Use the following command to restart the pods: 

oc -n istio-system delete pods -l 'app in (istiod,istio-ingressgateway)'.

The Operator will automatically recreate the pods after they have been deleted.

Step 4 (this is moved from the next section) - You must restart the bookinfo application pods so that the sidecar proxies pick up the secret changes.
Use the following command to restart the pods: 

oc -n bookinfo delete pods --all

You should see output similar to the following:

pod "details-v1-6cd699df8c-j54nh" deleted
pod "productpage-v1-5ddcb4b84f-mtmf2" deleted
pod "ratings-v1-bdbcc68bc-kmng4" deleted
pod "reviews-v1-754ddd7b6f-lqhsv" deleted
pod "reviews-v2-675679877f-q67r2" deleted
pod "reviews-v3-79d7549c7-c2gjs" deleted

The pods will be automatically recreated.  Verify that the pods are ready with the following command:

oc get pods -n bookinfo

After creating/changing/deleting the `cacert` secret, the control plane `istiod` and `gateway` pods must be restarted so the changes go into effect. Use the following command to restart the pods: `oc -n istio-system delete pods -l 'app in (istiod,istio-ingressgateway)'`.

Restart the application pods so that the sidecar containers pick up the changes as well.
====

[id="ossm-cert-manage-verify-cert_{context}"]
== Verifying your certificates

> Use the xref:../../service_mesh/v2x/ossm-create-mesh.adoc#ossm-tutorial-bookinfo-overview_ossm-create-mesh[Bookinfo example application] to verify that the workload certificates are signed by the certificates that were plugged into the CA. This requires you have `openssl` installed on your machine

Reviewer 4: ```suggestion
Use the xref:../../service_mesh/v2x/ossm-create-mesh.adoc#ossm-tutorial-bookinfo-overview_ossm-create-mesh[Bookinfo example application] to verify that the workload certificates are signed by the certificates that were plugged into the CA. 

.Prerequisites

* `openssl` is installed on your machine.

.Procedure

```

> . Extracting certificates from bookinfo workloads

Reviewer: This reads more like a heading than a step.  I would change this to "To extract certificates from bookinfo workloads use the following command:"

+

> Wait about 1 minute for the mTLS policty to take effect before retrieving the certificate chain of bookinfo. If Bookinfo was already deployed when you created the secret above, you need to restart the pods so they pick up the changes (see note above). Replace `<bookinfo-project>` with the project name where bookinfo workload is installed.

Reviewer: The "sleep 60" line in the command takes care of this 1 minute wait.
If you move the pod restart to the previous topic, that removes the need for second sentence AND the note.

If we just use `bookinfo`  instead of "replace <bookinfo-project> it increases usability.  Let's just assume that customers are following the directions and didn't install this anywhere with a different name.

+
[NOTE]
====
You must restart the bookinfo pods if they were already deployed. Use the following command to restart the pods: `oc -n <bookinfo-project> delete pods --all`.

If the pods are deployed after the secret changes made above, then you can skip this task.
====
+
[source,terminal]
----
$ sleep 60
$ oc -n <bookinfo-project> exec "$(oc -n <bookinfo-project> get pod -l app=productpage -o jsonpath={.items..metadata.name})" -c istio-proxy -- openssl s_client -showcerts -connect details:9080 > bookinfo-proxy-cert.txt
$ sed -n '/-----BEGIN CERTIFICATE-----/{:start /-----END CERTIFICATE-----/!{N;b start};/.*/p}' bookinfo-proxy-cert.txt > certs.pem
$ awk 'BEGIN {counter=0;} /BEGIN CERT/{counter++} { print > "proxy-cert-" counter ".pem"}' < certs.pem
----
+

> You should have three files in your working dir: `proxy-cert-1.pem`, `proxy-cert-2.pem` and `proxy-cert-3.pem`.

Reviewer: After running the command, you should have three files....  (Let's make it clear that this is a result)

Author: edit made

. Verify that the root certificate is the same as the one specified by the administrator. Replace `<path>` with the path to your certificates.
+
[source,terminal]
----
$ openssl x509 -in <path>/root-cert.pem -text -noout > /tmp/root-cert.crt.txt
$ openssl x509 -in ./proxy-cert-3.pem -text -noout > /tmp/pod-root-cert.crt.txt
$ diff -s /tmp/root-cert.crt.txt /tmp/pod-root-cert.crt.txt
----

> The /tmp/root-cert.crt.txt and /tmp/pod-root-cert.crt.txt flies are identical.

Reviewer: Let's make it clear that this is a result. I usually format the command line output as code.

You should see the following result:
The /tmp/root-cert.crt.txt and /tmp/pod-root-cert.crt.txt flies are identical.

Author: edit made

. Verify that the CA certificate is the same as the one specified by the administrator. Replace `<path>` with the path to your certificates.
+
[source,terminal]
----
$ openssl x509 -in ./proxy-cert-2.pem -text -noout > /tmp/pod-cert-chain-ca.crt.txt
$ diff -s /tmp/ca-cert.crt.txt /tmp/pod-cert-chain-ca.crt.txt
----

> The /tmp/ca-cert.crt.txt and /tmp/pod-cert-chain-ca.crt.txt files are identical.

Reviewer: And this was where I got stuck again.  Something isn't correct here.  I don't know if it's the command or formatting, but this is the point where things didn't work for me.

Reviewer: As discussed, I think we're missing a line in the comparison. From #36414

$ openssl x509 -in <path>/ca-cert.pem -text -noout > /tmp/ca-cert.crt.txt
$ openssl x509 -in ./proxy-cert-2.pem -text -noout > /tmp/pod-cert-chain-ca.crt.txt
$ diff -s /tmp/ca-cert.crt.txt /tmp/pod-cert-chain-ca.crt.txt

Result
Files /tmp/ca-cert.crt.txt and /tmp/pod-cert-chain-ca.crt.txt are identical

Author: edited these lines

. Verify the certificate chain from the root certificate to the workload certificate. Replace `<path>` with the path to your certificates.
+
[source,terminal]
----
$ openssl verify -CAfile <(cat <path>/ca-cert.pem <path>/root-cert.pem) ./proxy-cert-1.pem

> ./proxy-cert-1.pem: OK

Reviewer: I think this line is the output, not part of the command?

Author: Edit made

pod "ratings-v1-bdbcc68bc-kmng4" deleted
pod "reviews-v1-754ddd7b6f-lqhsv" deleted
pod "reviews-v2-675679877f-q67r2" deleted
pod "reviews-v3-79d7549c7-c2gjs" deleted
----

. Verify that the pods were created and are ready with the following command:
+

[source,terminal]
----
$ oc get pods -n bookinfo
----

[id="ossm-cert-manage-verify-cert_{context}"]
== Verifying your certificates

Use the Bookinfo sample application to verify that the workload certificates are signed by the certificates that were plugged into the CA. This requires you have `openssl` installed on your machine

. To extract certificates from bookinfo workloads use the following command:
+
[source,terminal]
----
$ sleep 60
$ oc -n <bookinfo-project> exec "$(oc -n <bookinfo-project> get pod -l app=productpage -o jsonpath={.items..metadata.name})" -c istio-proxy -- openssl s_client -showcerts -connect details:9080 > bookinfo-proxy-cert.txt
$ sed -n '/-----BEGIN CERTIFICATE-----/{:start /-----END CERTIFICATE-----/!{N;b start};/.*/p}' bookinfo-proxy-cert.txt > certs.pem
$ awk 'BEGIN {counter=0;} /BEGIN CERT/{counter++} { print > "proxy-cert-" counter ".pem"}' < certs.pem
----
+

> After running the command, you should have three files in your working directory: `proxy-cert-1.pem`, `proxy-cert-2.pem` and `proxy-cert-3.pem`.

Reviewer: "After running the command, you should have three files in your working directory: proxy-cert-1.pem, proxy-cert-2.pem and proxy-cert-3.pem."

I have bookinfo-proxy-cert.txt, certs,pem, proxy-cert-1.pem, proxy-cert-2.pem, so I'm not sure what's going on here.  Again, it's probably time to loop QE back in so we can get their help figuring what's going on here.
